### PR TITLE
Use callbacks in grid detection plan

### LINKED ISSUE #671: Use callbacks in grid detection
Currently the full grid plan passes in empty lists to grid detection plan that then are filled in with data to pass to the FGS. Instead this should be handled with callbacks.

Some of the callbacks will be ones that cover both detection and FGS (like the ispyb params). Some will bbe picking up data to then be passed to the FGS plan.

Note: We must retain the ability to run FGS independently by providing it expected values

### PR DESCRIPTION
**Pull request by noemifrisina**
Fixes #671 

Link to dodal PR (if required): [#91](https://github.com/DiamondLightSource/dodal/pull/91)
(remember to update `setup.cfg` with the dodal commit tag if you need it for tests to pass!)

### To test:
1.Run tests

### CODE REVIEW

File: src/artemis/experiment_plans/full_grid_scan.py [hunk 1]
```diff
@@ -90,23 +93,29 @@ def detect_grid_and_do_gridscan():
             f"{detector_params.prefix}_{detector_params.run_number}_{{angle}}"
         )
 
-        out_snapshot_filenames = []
-        out_upper_left = {}
+        oav_callback = OavSnapshotCallback()
+        yield from bps.subscribe(oav_callback)
 
         yield from grid_detection_plan(
```
Comment (DominicOram) [inline src/artemis/experiment_plans/full_grid_scan.py:99]: Should: I think you can do the subscribe/unsubscribe in a decorator

File: src/artemis/experiment_plans/oav_grid_detection_plan.py [hunk 2]
```diff
@@ -161,13 +149,13 @@ def grid_detection_main_plan(
...
         yield from bps.trigger(oav.snapshot, wait=True)
 
-        out_snapshot_filenames.append(
-            [
-                path_join(snapshot_dir, f"{snapshot_filename}_grid_overlay.png"),
-                path_join(snapshot_dir, f"{snapshot_filename}_outer_overlay.png"),
-                path_join(snapshot_dir, f"{snapshot_filename}.png"),
-            ]
-        )
+        yield from bps.create("snapshot_directories")
```
Comment (DominicOram) [inline src/artemis/experiment_plans/oav_grid_detection_plan.py:152]: Should: This isn't just directories any more

**Review by DominicOram — CHANGES_REQUESTED** [review]
Minor comments in code, I think we should also merge https://github.com/DiamondLightSource/python-artemis/pull/730 first so that we can ensure it doesn't break tests

**Comment by noemifrisina** [comment]
Done, thanks

**Comment by DominicOram** [comment]
Could you:
* Fix conflicts
* Confirm test in https://github.com/DiamondLightSource/python-artemis/pull/730 still works on this
* Add new tests if you think that https://github.com/DiamondLightSource/python-artemis/pull/730 doesn't fully test the functionality

**Review by DominicOram — APPROVED** [review]
Looks good, thanks

**Pull request merged**